### ISSUE TITLE
Error logging messages when exceptions occurs (probably specific to async contexts)

### ISSUE THREAD

**Issue by luanmm** [microsoft/node-api-dotnet#194]
Hello, dear maintainers.

I'm getting the same errors everytime some exception throws (on .NET side or in script side).

```sh
      System.AggregateException: An error occurred while writing to logger(s). (There is no active JS value scope.
      Current thread: #29 ".NET TP Worker". Consider using the synchronization context to switch to the JS thread.)
       ---> Microsoft.JavaScript.NodeApi.JSInvalidThreadAccessException: There is no active JS value scope.
      Current thread: #29 ".NET TP Worker". Consider using the synchronization context to switch to the JS thread.
         at Microsoft.JavaScript.NodeApi.JSValueScope.get_Current()
         at Microsoft.JavaScript.NodeApi.JSValueScope.get_CurrentRuntime()
         at Microsoft.JavaScript.NodeApi.JSReference.GetValue()
         at Microsoft.JavaScript.NodeApi.JSError.get_Value()
         at Microsoft.JavaScript.NodeApi.JSException.get_StackTrace()
         at System.Exception.ToString()
         at Microsoft.Extensions.Logging.Console.SimpleConsoleFormatter.CreateDefaultLogMessage[TState](TextWriter textWriter, LogEntry`1& logEntry, String message, IExternalScopeProvider scopeProvider)
         at Microsoft.Extensions.Logging.Console.SimpleConsoleFormatter.Write[TState](LogEntry`1& logEntry, IExternalScopeProvider scopeProvider, TextWriter textWriter)
         at Microsoft.Extensions.Logging.Console.ConsoleLogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
         at Microsoft.Extensions.Logging.Logger.<Log>g__LoggerLog|14_0[TState](LogLevel logLevel, EventId eventId, ILogger logger, Exception exception, Func`3 formatter, List`1& exceptions, TState& state)
         --- End of inner exception stack trace ---
         at Microsoft.Extensions.Logging.Logger.ThrowLoggingError(List`1 exceptions)
         at Microsoft.Extensions.Logging.Logger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
         at Microsoft.Extensions.Logging.LoggerMessage.<>c__DisplayClass12_0`2.<Define>g__Log|0(ILogger logger, T1 arg1, T2 arg2, Exception exception)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
```

Seems to be something regarding the synchronization context between threads that stops the logging functionality.

I'm using promises/tasks in my script, like below:

```csharp
        await nodeEnv.RunAsync(async () =>
        {
            var script = nodeEnv.Import(scriptPath);

            var asyncFn = JSValue.CreateFunction("asyncFn", (args) =>
            {
                var data = (NodeStream)args[0];

                return new JSPromise(async (resolve) =>
                {
                    var buffer = new byte[4096].AsMemory();
                    int count;

                    using var stream = new MemoryStream();

                    while ((count = await data.ReadAsync(buffer)) > 0)
                    {
                        await stream.WriteAsync(buffer[..count]);
                    }

                    // TODO: Save stream to file

                    resolve(JSValue.Undefined);
                });
            });

            var context = JSValue.CreateObject();
            context.SetProperty("asyncFn", asyncFn); // This "asyncFn" is called in the "main" method (in the script side)

            var result = script.CallMethod("main", context);
            await JSPromise.Resolve(result).AsTask();
        });
```

This is blocking a lot any possibility of troubleshooting errors, because the only message I'm getting is the stated above.

If needed, I can help on this (with some guidance).

Thanks in advance for any thoughts/help.

**Comment by luanmm** [comment]
Updates:

I could check the error putting a try/catch in the "main" block (in the scripting side), for now, as workaround:

```js
export const main = async (context) => {
  try {
     // ...
        await context.asyncFn(/*...*/);
     // ...
  }
  catch (error) {
     console.error("Error", error);
  }
};
```

This allows me to troubleshoot, but I think this try/catch should be optionally anyway. The error should propagate back to .NET side to give the entire stack, right? At least, this is something that is really great [on the libraries proposal](https://github.com/microsoft/node-api-dotnet?tab=readme-ov-file#error-propagation).

**Comment by jasongin** [comment]
Thanks for reporting this. I think this should be handled better by the library.

The `JSInvalidThreadAccessException` was thrown because the `JSException.StackTrace` property was trying to access the JS environment (a reference to a JS `Error` object) from a thread other than the JS main thread. As you say, that caused the original exception message and stack trace to be lost, which is very unhelpful. The original `JSException` was thrown on the JS thread, but then it escaped via the `RunAsync()` task failure out to the ASP.NET thread, while still holding a reference to that JS `Error` value.

I think there could be two approaches to fixing this problem in `JSException`:
  1. When constructing a `JSException` from a JS `Error` value, immediately format the Error's JS stack trace as a string and save that in a member, instead of holding a reference to the Error until the stack trace is requested. This would be an easy change, but would make it expensive when JS errors/exceptions are thrown and caught without ever requesting the stack trace.
  2. Convert the JS stack trace to a string only when returning a `JSException` via a failed task from `RunAsync()` or similar methods which are likely to be called from a non-JS thread.

Meanwhile, another workaround which you may like better could be to add an exception-handler like this to your C# code:

```C#
await nodeEnv.RunAsync(async () =>
{
    try
    {
        ...
    }
    catch (JSException ex)
    {
        throw new Exception("Exception thrown from JS environment: " + ex.ToString());
    }
});
```

The `ex.ToString()` there will cause the JS stack trace to be converted to a string before it's passed back to the calling thread.

**Comment by jasongin** [comment]
Fixed in version 0.6.
